### ISSUE TITLE
Epic: harden the SkillOpt evaluator — measure it, lean on hard verifiers, prefer pairwise

### ISSUE THREAD

**Issue by jerryfane** [gitmoot/gitmoot#344]
> **Tracking epic — not a work item.** Sub-issues hold the work; this frames the thesis and sequence.

## Thesis

A research session + a code-grounded assessment of `gitmoot-skillopt` surfaced a coherent way to harden the LLM **judge/evaluator** that drives prompt optimization. The key reframe:

**The LLM judge is not the only safety net.** A deterministic hard-verifier floor already short-circuits before the judge (`evaluator.py:195-261`), and **promotion is manual**. So the real failure modes are not "a gamed judge auto-ships garbage" — they are:

1. **Low judge precision** → wasted human review time, and
2. **A blind-spotty judge** → a weak optimization signal (the optimizer only "sees" human intent *through* the judge's structured fields).

The literature (LLM-as-a-Judge surveys, reward-model overoptimization, juries, process reward models) points the same way. Priority order:

```mermaid
flowchart LR
  M["1. MEASURE the judge<br/>(judge↔human disagreement)"] --> H["2. HARD verifiers<br/>(move checkable things out of the LLM)"]
  H --> P["3. PREFER pairwise<br/>(#77 live A/B)"]
  P --> S["4. Judge SEES MORE<br/>(rubric induction · trajectory digest)"]
  S --> J["5. Judge JURY<br/>(escalation / promotion-boundary)"]
  M -. feeds judge-prompt opt .-> S
  M -. feeds judge-prompt opt .-> J
```

## Design constraints (all sub-issues respect these)

- **No model-weight training.** Calibrate via captured data, few-shot anchors, or a light regression layer — never fine-tune the judge.
- **Additive to `contract_version=1`** (`internal/skillopt/contract.go:21-25`). Prefer the documented opaque passthroughs (`evaluator_config`, `FeedbackContext`, `eval_artifacts`, item/option `MetadataJSON`) over new typed contract fields for first-cut work.
- **Local-first; manual promotion stays.**

## Sub-issues

1. ~~**#345** — Capture judge↔human outcomes → optimize the judge prompt (per task-kind)~~ — ⭐ keystone — ✅ **done** (Phase 1 + Phase 2 shipped; production-apply tracked in #354)
2. ~~**#346** — Extend programmatic hard verifiers (shrink what the judge guesses)~~ — do-first floor — ✅ **done** (gitmoot-skillopt#73; wheel release pending)
3. **#347** — Offline rubric/metric induction from human feedback (AutoLibra-style)
4. **#348** — Give the judge a budgeted trajectory digest (process, not just artifact)
5. **#349** — Judge jury via the `review-panel` recipe (diverse-judge ensemble)
6. **#77** — Live pairwise evaluation mode *(existing)* — the "prefer pairwise" lever; substrate that makes #345 & #348 faithful

## Sequencing note (an honest dissent)

The two independent research agents (codex, kimi) + this analysis agree on the *set* and that **capture (#1)** and **hard verifiers (#2)** come first — but disagree on which is #1: codex leads with **capture** (it makes the judge *measurable*, unlocking everything else); kimi leads with **hard verifiers** (cheapest deterministic floor, reduces judge load). They are parallelizable and both are "do first." Everything after depends on the measurement loop (#1) existing.

## Provenance

Implementation proposals + citations below are a **combination** of (a) the grounded code reading of both repos, (b) a web SOTA sweep, and (c) two independent gitmoot-orchestra research agents (codex + kimi). Dissent is noted where it exists.

**Comment by jerryfane** [comment]
**Epic progress (2026-06-20)**

- **#345 — keystone (capture → optimize the judge prompt): ✅ implemented + merged.** gitmoot #350 (capture + `judge-report`), #351 (per-`task_kind` judge-prompt contract carry), #352 (real landing-page report-shape fix), #353 (judge-report calibration fix); gitmoot-skillopt jerryfane/gitmoot-skillopt#70 (per-`task_kind` judge prompt, `human_agreement` gate, freeze-and-alternate tuning). The **Phase-1 capture path is now validated by a full live codex E2E** (real `codex_exec` target + real codex LLM judge → all four judge↔human directions captured through the real binary → `judge-report` matrix/κ/calibration). Phase-2 optimization real-data validation remains deferred until captured outcomes accumulate. See #345 for details.
- **#346** (extend programmatic hard verifiers) — not started.
- **#347** (offline AutoLibra-style rubric induction → `evaluator_config['rubric']`) — not started.
- **#348** (budgeted trajectory digest to the judge; durable `trajectories[]`) — not started.
- **#349** (judge jury via `review-panel`; cross-family ensemble + minority veto) — not started.
- **#77** (live pairwise eval — the "prefer pairwise" lever) — commented with bias mitigations; not started.

The measurement substrate (#345) being live now unblocks measuring the impact of #346–#349 against real human-agreement data.

**Comment by jerryfane** [comment]
**Status (2026-06-27): all actionable sub-issues complete.**

- 1. Measure — #345 ✅
- 2. Hard verifiers — #346 ✅
- 3. Prefer pairwise — #77 ✅ (split + delivered: #507 fork producer, #508 Go ingest)
- 4. Judge sees more — #348 **Phase 1 ✅** (gitmoot-skillopt#75, trajectory digest); Phase 2 (durable `trajectories[]`) deferred by design, gated on Phase 1 showing verdict deltas via #345 data — #348 stays open
- 5. Judge jury — #349 ✅ (#512)

**Remaining (both blocked on a prerequisite, not effort):** #347 (rubric induction — needs an accumulated human-feedback corpus) and #348 Phase 2 (gated on Phase 1 evidence). Keeping this epic open to track those two.

**Comment by jerryfane** [comment]
Adding **Reliability without Validity** as a supporting reference for this evaluator-hardening epic.

Paper: [Reliability without Validity: A Systematic, Large-Scale Evaluation of LLM-as-a-Judge Models Across Agreement, Consistency, and Bias](https://arxiv.org/abs/2606.19544)  
Local refs from the research pass: `references/2606.19544.pdf`, `references/2606.19544.md`

The simple takeaway is: an LLM judge is not trustworthy just because it agrees often or gives stable answers. It can be consistently biased. For example, a judge can repeatedly prefer answer A because it appears first; that looks stable under test-retest, but it is not valid.

This reinforces the direction of #344, especially the "measure the judge" part. Gitmoot already has the right foundation with judge↔human outcome capture, `skillopt judge-report`, Cohen's kappa, hard verifiers, and manual promotion. The paper sharpens what the validation checklist should include before we trust a judge as an optimization signal.

Gitmoot implications:

1. **Make kappa the headline metric**
   Raw agreement should be secondary. The paper shows exact-match agreement can overstate judge quality by tens of percentage points because it does not correct for chance.

2. **Add position-swap audits for pairwise reviews**
   For A/B SkillOpt reviews, evaluate both A/B and B/A on a sample and report position bias as `|P(A wins) - 0.5|`. A judge that changes too much when order changes should be down-weighted or treated as untrusted.

3. **Run repeated judge checks on a small validation slice**
   The paper recommends at least 3 independent runs at temperature 0 with response caching disabled. Gitmoot does not need to do this for every evaluation, but it would be useful for `judge-report` / judge-prompt validation.

4. **Do not trust consistency alone**
   High test-retest reliability is only good if position bias is low. If a judge is stable but biased, that is a failure mode, not a strength.

5. **Validate by task kind**
   Judge quality does not transfer cleanly across benchmarks. For Gitmoot this means judge reports should stay scoped by `task_kind` where possible: code correctness, design review, scope fidelity, docs quality, etc.

6. **Keep hard verifiers and human promotion as the floor**
   The paper does not mean "do not use LLM judges." It means LLM judges should be measured, chance-corrected, bias-audited, and treated as one signal below hard checks and human decisions.

Related local doc points:

- `gitmoot skillopt judge-report` already reports agreement and Cohen's kappa.
- #525/#527 cover finer-grained binary checks and prompt-update signals.
- #348 covers giving judges more process signal.
- This paper mainly suggests adding the missing bias/replicate parts to the existing #344 direction rather than opening a separate broad issue.

**Comment by jerryfane** [comment]
Epic progress (2026-07-02). 1) Measure the judge - done: #345 (skillopt capture + per-task_kind judge prompt) plus the gitmoot-side 'gitmoot skillopt judge agreement' (raw %, Cohen's kappa, per-dimension, --json) merged #599 (v0.8.2); the live Mode A loop (real merged-PR + CI outcomes -> optimizer -> better candidate) was proven end-to-end and closed (#538). 2) Hard verifiers - done: #346 (gitmoot-skillopt#73); a deeper deterministic tier RFC #474 (fresh-sandbox exit-0 grader, slime-informed) is open and now folded under this epic. 3) Prefer pairwise (#77) - done: split + delivered via #507 (fork dual-rollout producer) + #508 (Go ingest). 4) Judge sees more (#348) - Phase 1 done (gitmoot-skillopt#75 trajectory digest); Phase 2 (durable trajectories[]) deferred, gated on Phase 1 verdict-delta evidence via #345. 5) Judge jury (#349) - done: #512 (cross-family ensemble). Remaining, keeping this open: #347 (rubric induction - blocked on an accumulated human-feedback corpus), #348 Phase 2 (gated), #474 (hard-verifier tier RFC), and the bias/replicate additions surfaced by the 'Reliability without Validity' comment (kappa-as-headline, pairwise position-swap audits, >=3-run temp-0 replicate checks) tracked via #525/#527." Also update the body's Sub-issues list so #349, #77 (#507/#508) are struck through as done and #474 is listed under hard verifiers.

**Comment by jerryfane** [comment]
Closing the epic — every track of the measure → hard-verifiers → pairwise → jury sequence has now shipped and been verified:

- **Measure the judge**: #345 (keystone) + the judge↔human agreement harness (`skillopt judge agreement`, per-comparison join, Cohen's kappa with small-N honesty) in #599; the judge is now a *measured* instrument, not an assumed one.
- **Hard verifiers**: #346's static deterministic checkers (#485), completed by the fresh-sandbox exit-0 **hard-verifier tier** in #622/#474 — independent `git clone --local` sandboxes (no shared git state, severed origin), tri-state results so setup failures can't fabricate training negatives, and the fabricated 0.5 merged-no-CI score replaced with honestly-grounded bands (composing with #614's confirmed-no-CI gate status).
- **Pairwise**: #77 → #507/#508 (shipped, incl. the cross-repo round-trip E2E #514).
- **Jury**: #349 → #512 (cross-family jury with veto dimensions).

The full loop was also proven live end-to-end (#538: real PR + CI outcomes → Mode A harvest → optimizer candidate). Remaining research seeds (#347 rubric induction, #348 phase 2, #525-#527 BINEVAL) are tracked in their own issues and are increments *on top of* this foundation, not part of the epic's thesis. 🎉
